src/internal/orm/tx.go

```diff
@@ -0,0 +1,90 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orm
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/astaxie/beego/orm"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+var randNumberString = func(n int) string {
+	letterBytes := []byte("1234567890")
+
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+
+	return string(b)
+}
+
+var getTimestamp = func() int64 {
+	return time.Now().Unix()
+}
+
+func getSavepoint() string {
+	return fmt.Sprintf("p%dr%s", getTimestamp(), randNumberString(5))
+}
+
+// ormerTx transaction which support savepoint
+type ormerTx struct {
+	orm.Ormer
+	savepoint string
+}
+
+func (o *ormerTx) Begin() error {
+	err := o.Ormer.Begin()
+	if err != nil && err == orm.ErrTxHasBegan {
```

Comment:
if err != nil && err == orm.ErrTxHasBegan =>

`if err := o.Ormer.Begin(); err == orm.ErrTxHasBegan { ... }` is enough

Reply:
This is not enough, the func will return the err when it is not `ErrTxHasBegan`.

Reply:
Move operations about savepoint to methods which made this more more simpler.